client/client.go

```diff
@@ -283,138 +287,113 @@ func (c *client) tsLoop() {
-					case tsDeadlineCh <- dl:
-					case <-loopCtx.Done():
-						cancel()
-						return
-					}
-					opts := extractSpanReference(reqs)
-					if processErr := c.processTSORequests(streams[dc], dc, reqs, opts); processErr != nil {
-						err.Store(processErr)
-					}
-					close(done)
-				}(dcLocation, requests)
-			}
-			wg.Wait()
 		case <-loopCtx.Done():
-			cancel()
 			return
+		default:
 		}
-		if err.Load() != nil {
-			select {
-			case <-loopCtx.Done():
-				cancel()
-				return
-			default:
+		for _, dcLocation := range c.getDCLocations() {
+			if !c.checkTSODispatcher(dcLocation) {
+				c.createTSODispatcher(dcLocation)
+				go func(dc string, tsoDispatcher chan *tsoRequest) {
+					var (
+						err       error
+						ctx       context.Context
+						cancel    context.CancelFunc
+						stream    pdpb.PD_TsoClient
+						opts      []opentracing.StartSpanOption
+						createdCh = make(chan struct{})
+						requests  = make([]*tsoRequest, maxMergeTSORequests+1)
+					)
+					for {
+						if stream == nil {
+							ctx, cancel = context.WithCancel(loopCtx)
+							go c.checkStreamTimeout(ctx, cancel, createdCh)
+							stream, err = pdpb.NewPDClient(c.connMu.clientConns[c.connMu.allocators[dc]]).Tso(ctx)
+							if err != nil {
+								select {
+								case <-ctx.Done():
+									return
+								default:
+								}
+								log.Error("[pd] create tso stream error", zap.String("dc-location", dc), errs.ZapError(errs.ErrClientCreateTSOStream, err))
+								c.ScheduleCheckLeader()
+								cancel()
+								c.revokeTSORequest(errors.WithStack(err), tsoDispatcher)
+								select {
+								case <-time.After(time.Second):
+								case <-ctx.Done():
+									return
+								}
+								continue
+							}
+							if stream != nil {
+								createdCh <- struct{}{}
+							}
+						}
+						select {
+						case first := <-tsoDispatcher:
+							pendingPlus1 := len(tsoDispatcher) + 1
+							requests[0] = first
+							for i := 1; i < pendingPlus1; i++ {
+								requests[i] = <-tsoDispatcher
+							}
+							done := make(chan struct{})
+							dl := deadline{
+								timer:  time.After(c.timeout),
+								done:   done,
+								cancel: cancel,
+							}
+							c.tsDeadline.RLock()
+							tsDeadlineCh := c.tsDeadline.ch[dc]
+							c.tsDeadline.RUnlock()
+							select {
+							case tsDeadlineCh <- dl:
+							case <-ctx.Done():
+								return
+							}
+							opts = extractSpanReference(requests[:pendingPlus1], opts[:0])
+							err = c.processTSORequests(stream, dc, requests[:pendingPlus1], opts)
+							close(done)
+						case <-ctx.Done():
+							return
+						}
+						if err != nil {
+							select {
+							case <-loopCtx.Done():
+								return
+							default:
+							}
+							log.Error("[pd] getTS error", zap.String("dc-location", dc), errs.ZapError(errs.ErrClientGetTSO, err))
+							c.ScheduleCheckLeader()
+							cancel()
+							stream = nil
+						}
+					}
+				}(dcLocation, c.tsoDispatcher.ch[dcLocation])
```

Comment:
Does the outer loop need to wait the inner loop's goroutine?

Reply:
The outer loop is used to create the goroutine for a new `dc-location`, so it doesn't need to wait for the inner goroutines. The `loopCtx` controls the lifetimes of the inner goroutines.

Reply:
Does `tsLoop` always run the loop started in `line 290`?

Reply:
Yes, it will keep looping to check whether we have new DC joined.